services-core/src/solution_submission/gas_price_increase.rs

```diff
@@ -0,0 +1,77 @@
+use futures::stream::{Stream, StreamExt};
+
+// openethereum requires that the gas price of the resubmitted transaction has increased by at
+// least 12.5%.
+const MIN_GAS_PRICE_INCREASE_FACTOR: f64 = 1.125 * (1.0 + f64::EPSILON);
+
+pub fn minimum_increase(gas_price: f64) -> f64 {
+    (gas_price * MIN_GAS_PRICE_INCREASE_FACTOR).ceil()
+}
+
+fn new_gas_price_estimate(
+    previous_gas_price: f64,
+    new_gas_price: f64,
+    max_gas_price: f64,
+) -> Option<f64> {
+    let min_gas_price = minimum_increase(previous_gas_price);
+    if min_gas_price > max_gas_price {
+        return None;
+    }
+    if new_gas_price <= previous_gas_price {
+        // Gas price has not increased.
+        return None;
+    }
+    // Gas price could have increased but doesn't respect minimum increase so adjust it up.
+    let new_price = min_gas_price.max(new_gas_price);
+    Some(new_price.min(max_gas_price))
+}
+
+// Adapt a stream of gas prices to only yield gas prices that respect the minimum gas price increase.
```

Comment:
nit: Please add to the comment that prices are capped at a maximum (I think it could even go in the method name as `enforce_minimum_increase_up_to` or even be implemented as a separate chained combinator, but not important).